sunbeam-python/sunbeam/core/watcher.py

```diff
@@ -0,0 +1,175 @@
+        service_type="infra-optim",
+        region_name=region,
+    )
+    return watcher_client.Client(session=conn.session, endpoint=watcher_endpoint)
+
+
+def _create_host_maintenance_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    template = client.audit_template.create(
+        name=ENABLE_MAINTENANCE_AUDIT_TEMPLATE_NAME,
+        description="Audit template for cluster maintaining",
+        goal=ENABLE_MAINTENANCE_GOAL_NAME,
+        strategy=ENABLE_MAINTENANCE_STRATEGY_NAME,
+    )
+    return template
+
+
+def _create_workload_balancing_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    template = client.audit_template.create(
+        name=WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME,
+        description="Audit template for workload balancing",
+        goal=WORKLOAD_BALANCING_GOAL_NAME,
+        strategy=WORKLOAD_BALANCING_STRATEGY_NAME,
+    )
+    return template
+
+
+def get_enable_maintenance_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    try:
+        template = client.audit_template.get(ENABLE_MAINTENANCE_AUDIT_TEMPLATE_NAME)
+    except NotFound:
+        template = _create_host_maintenance_audit_template(client=client)
+    return template
+
+
+def get_workload_balancing_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    try:
+        template = client.audit_template.get(WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME)
+    except NotFound:
+        template = _create_workload_balancing_audit_template(client=client)
+    return template
+
+
+@timeout_decorator.timeout(TIMEOUT)
+def create_audit(
+    client: watcher_client.Client,
+    template: watcher.AuditTemplate,
+    audit_type: str = "ONESHOT",
+    parameters: dict[str, Any] = {},
+) -> watcher.Audit:
+    audit = client.audit.create(
+        audit_template_uuid=template.uuid,
+        audit_type=audit_type,
+        parameters=parameters,
+    )
+    while True:
+        audit_details = client.audit.get(audit.uuid)
+        if audit_details.state in ["SUCCEEDED", "FAILED"]:
+            break
+        time.sleep(SLEEP_INTERVAL)
+    if audit_details.state == "SUCCEEDED":
+        LOG.debug(f"Create Watcher audit {audit.uuid} successfully")
+    else:
+        LOG.debug(f"Create Watcher audit {audit.uuid} failed")
+        raise SunbeamException(
+            f"Create watcher audit failed, template: {template.name}"
+        )
+
+    _check_audit_plans_recommended(client=client, audit=audit)
+    return audit
+
+
+def _check_audit_plans_recommended(client: watcher_client.Client, audit: watcher.Audit):
+    action_plans = client.action_plan.list(audit=audit.uuid)
+    # Verify all the action_plan's state is RECOMMENDED
+    if not all(plan.state in ["RECOMMENDED", "SUCCEEDED"] for plan in action_plans):
+        raise SunbeamException(
+            f"Not all action plan for audit({audit.uuid}) is RECOMMENDED"
+        )
```

Comment:
The comments and message says recommended, but the code actually checks recommended OR succeeded. Is this intentional?

Reply:
So, there are two different situations:

- State in recommended: Action plan generated, pending action.
- State in succeeded: Action plan generated, no action created.

This is to avoid to stuck/race exception in the second situation.

Reply:
Updated.